### ISSUE TITLE
Enable .doo files to scale to large libraries

### ISSUE THREAD

**Issue by keyboardDrummer** [dafny-lang/dafny#4721]
Currently Dafny .doo files always contain all the Dafny source code that was used as part of the Dafny compilation, including the source of libraries.

This does not scale when libraries and their dependencies get larger.

Dafny should support `.doo` files that only contain the interface of the library that they were built form, where the interface contains all the method and function signatures including their verification related parts.

The interface `interface.dfy` inside `library.doo` should be used when using the `.doo` file as a library using `--library=library.doo`

When using a `.doo` file as source, using `dafny <command> source.dfy library.doo`, Dafny should throw an error when the .doo file only contains an interface, but no sources.

**Comment by robin-aws** [comment]
> Currently Dafny .doo files always contain all the Dafny source code that was used as part of the Dafny compilation, including the source of libraries.

Not necessarily! If you pass the `.doo` file as a source file, as in `dafny consumer.dfy library.doo`, then yes, it will be compiled as well, whether to a target language or into a `consumer.doo` file. But if you pass it as a library file, as in `dafny consumer.dfy --library library.doo`, then it won't be included.

We're using the former approach for `DafnyStandardLibraries.doo` only because using `--library` instead and building it into the runtimes was blocked by #511. Once the fix for that is in I'm planning to correct that.

It's true that if you build `B.doo` using `--library A.doo`, `B.doo` won't contain any definitions of the contents of `A`, interface or otherwise. That does mean that if you want `C` to use `B`, you also have to get a hold of the transitive dependency `--library A.doo` since we have to resolve the entire program. That's not necessarily a big problem though, because you should be able to rely on a packaging solution to do that for you: the Dafny Gradle plugin will collect all transitive `.doo` files for you based on the existing Gradle dependency management support.

It is still probably worth figuring out a way to not have to re-resolve the whole program graph though, for efficiency if not necessarily to avoid having to build with all transitive dependencies.

**Comment by keyboardDrummer** [comment]
> Not necessarily! If you pass the .doo file as a source file, as in dafny consumer.dfy library.doo, then yes, it will be compiled as well, whether to a target language or into a consumer.doo file. But if you pass it as a library file, as in dafny consumer.dfy --library library.doo, then it won't be included.

I meant that when you do `dafny build --target=lib source1.dfy source2.dfy --library=lib1.dfy,lib2.doo`, then everything from `source1.dfy`, `source2.dfy`,`lib1.dfy` and `lib2.doo` will be included in the generated `.doo` file

**Comment by robin-aws** [comment]
> I meant that when you do dafny build --target=lib source1.dfy source2.dfy --library=lib1.dfy,lib2.doo, then everything from source1.dfy, source2.dfy,lib1.dfy and lib2.doo will be included in the generated .doo file

If you're seeing that somehow it's a bug. If you run `lit` on the [separate-verification/app.dfy test](https://github.com/dafny-lang/dafny/blob/master/Source/IntegrationTests/TestFiles/LitTests/LitTest/separate-verification/app.dfy) (just because lit dumps the files in a more convenient place) and then unzip `Inputs/seq.doo`, you'll see the `program` file inside it does not include the contents of `wrappers.doo`.

The code snippet you shared ends up not printing the contents of `--library` files because of this: https://github.com/dafny-lang/dafny/blob/master/Source/DafnyCore/AST/Grammar/Printer.cs#L393-L398 :)
